allennlp/models/coreference_resolution/coref.py

```diff
@@ -552,6 +566,66 @@ def forward(self,  # type: ignore
             output_dict["loss"] = loss
         return output_dict
 
+    def decode(self, output_dict: Dict[str, torch.Tensor]):
+        """
+        Converts the list of spans and predicted antecedent indices into clusters
+        of spans for each element in the batch.
+
+        Parameters
+        ----------
+        output_dict : ``Dict[str, torch.Tensor]``, required.
+            The result of calling :func:`~forward` on an instance or batch of instances.
+
+        Returns
+        -------
+        The same output dictionary, but with an additional ``clusters`` key:
+
+        clusters : ``List[List[List[Tuple[int, int]]]]``
+            A nested list, representing, for each batch, the list of clusters, which are
+            in turn comprised of a list of (start, end) inclusive spans into the original
+            document.
+        """
+        top_spans = output_dict["top_spans"].data.cpu()
+        predicted_antecedents = output_dict["predicted_antecedents"].data.cpu()
+        all_clusters: List[List[List[Tuple[int, int]]]] = []
+
+        # Calling zip() on two tensors results in an iterator over their
+        # first dimension. This is  iterating over instances in the batch.
+        for all_spans, span_antecedents in zip(top_spans, predicted_antecedents):
+
+            clusters: Dict[int, List[Tuple[int, int]]] = {}
+            for span, antecedent_index in zip(all_spans, span_antecedents):
+                if antecedent_index != -1:
+                    # Find the right cluster to update with this span.
+                    # We might have referred to a span which in turn
+                    # refers to some previous span, so here we trace back
+                    # through the list to find the "root" of the cluster,
+                    # and only update that set.
+                    cluster_index_to_update = antecedent_index
```

Comment:
What you have is fine, but this might be slightly simpler and need less explanation if you do:

```python
cluster = clusters.setdefault(antecedent_index, set([]))
clusters[current_span_index] = cluster
cluster.add(antecedent_span)
cluster.add(current_span)
```

This way you're just making the `clusters[span]` object refer to the same set for all coreferent indices, and you always add both the antecedent span and the current span, because it's a set.

Hmm, this might break the call to `.values()` at the end, though.  I guess you could add those to a set, too, then return the result as a list...

Reply:
Yeah I originally started out with a `defaultdict(set)`, but then I thought this way was more explicit about how we have to remember to add the root span - i've changed it back.